From 1dbe962fc1b4e64473388026cdc6669a71264208 Mon Sep 17 00:00:00 2001
From: Nikita Chernikov <nchernikov@maxlinear.com>
Date: Thu, 14 May 2026 14:08:56 +0300
Subject: [PATCH] WLANRTSYS-99073 Add RSN Override elements to MBSSID non-TX
 BSS profiles when not inherited from TX BSS

Fixup to: WLANRTSYS-63078 Multi BSS changes for MLO

Signed-off-by: Nikita Chernikov <nchernikov@maxlinear.com>
---
 src/ap/ieee802_11_shared.c | 15 ++++++++++++---
 src/ap/mxl_hostapd.c       |  7 +++++++
 src/ap/wpa_auth_ie.c       | 33 ++++++++++++++++++++++++++++++---
 src/common/wpa_common.h    |  3 +++
 4 files changed, 52 insertions(+), 6 deletions(-)

diff --git a/src/ap/ieee802_11_shared.c b/src/ap/ieee802_11_shared.c
index 043a26549..46883d305 100644
--- a/src/ap/ieee802_11_shared.c
+++ b/src/ap/ieee802_11_shared.c
@@ -1466,7 +1466,13 @@ static bool is_security_ie_required(struct hostapd_data *nontx_bss, struct hosta
 		(nontx_bss->conf->group_cipher != tx_bss->conf->group_cipher) ||
 		(nontx_bss->conf->rsn_preauth != tx_bss->conf->rsn_preauth) ||
 		(nontx_bss->conf->group_mgmt_cipher != tx_bss->conf->group_mgmt_cipher) ||
-		(nontx_bss->conf->ieee80211w != tx_bss->conf->ieee80211w))) {
+		(nontx_bss->conf->ieee80211w != tx_bss->conf->ieee80211w) ||
+		(nontx_bss->conf->rsn_override_key_mgmt != tx_bss->conf->rsn_override_key_mgmt) ||
+		(nontx_bss->conf->rsn_override_key_mgmt_2 != tx_bss->conf->rsn_override_key_mgmt_2) ||
+		(nontx_bss->conf->rsn_override_pairwise != tx_bss->conf->rsn_override_pairwise) ||
+		(nontx_bss->conf->rsn_override_pairwise_2 != tx_bss->conf->rsn_override_pairwise_2) ||
+		(nontx_bss->conf->rsn_override_mfp != tx_bss->conf->rsn_override_mfp) ||
+		(nontx_bss->conf->rsn_override_mfp_2 != tx_bss->conf->rsn_override_mfp_2))) {
 		nontx_bss->conf->wpa_ie_required_for_mbss = 1;
 		return true;
 	}
@@ -1500,6 +1506,9 @@ static int get_non_tx_bss_len(struct hostapd_data *bss, bool is_beacon,
 
 	if (is_security_ie_required(bss, bss->iface->bss[tx_bss_idx])) {
 		len = len + WPA_IE_MAX_LEN;
+		if (bss->conf->rsn_override_key_mgmt ||
+		    bss->conf->rsn_override_key_mgmt_2)
+			len = len + WPA_RSNO_IE_EXTRA_LEN;
 	}
 
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
@@ -1641,7 +1650,7 @@ u8 * hostapd_eid_multiple_bssid(struct hostapd_data *hapd, u8 *eid, bool is_beac
 #endif
 
 
-	wpa_ie = os_malloc(WPA_IE_MAX_LEN);
+	wpa_ie = os_malloc(WPA_IE_MAX_LEN + WPA_RSNO_IE_EXTRA_LEN);
 	if (wpa_ie == NULL) {
 		wpa_printf(MSG_ERROR,"%s memory allocation failed for wpa_ie",__func__);
 		return NULL;
@@ -1697,7 +1706,7 @@ u8 * hostapd_eid_multiple_bssid(struct hostapd_data *hapd, u8 *eid, bool is_beac
 		}
 		wpa_ie_len = 0;
 
-		os_memset(wpa_ie, 0, WPA_IE_MAX_LEN);
+		os_memset(wpa_ie, 0, WPA_IE_MAX_LEN + WPA_RSNO_IE_EXTRA_LEN);
 
 		*eid++ = WLAN_EID_SUBELEMENT_NONTRANSMITTED_BSSID_PROFILE;
 		subelem_size_pos = eid;
diff --git a/src/ap/mxl_hostapd.c b/src/ap/mxl_hostapd.c
index 017354039..26dc44d10 100644
--- a/src/ap/mxl_hostapd.c
+++ b/src/ap/mxl_hostapd.c
@@ -3281,6 +3281,13 @@ bool mxl_hostapd_check_non_inherit_rsnxe(struct hostapd_data *bss)
 	size_t tx_bss_idx = mxl_hostapd_get_mbssid_tx_bss_idx(bss->iface);
 	struct hostapd_data *hapd = bss->iface->bss[tx_bss_idx];
 
+	/* WPA3-Personal Compatibility Mode: RSNXE is omitted from beacon,
+	 * RSNXE Override is used instead. Non-inherit RSNXE from TX BSS.
+	 * Check this before TX BSS SAE check since Compatibility Mode
+	 * may use rsn_override_key_mgmt for SAE instead of wpa_key_mgmt. */
+	if (bss->conf->rsn_override_omit_rsnxe)
+		return true;
+
 	if (!(hapd->conf->wpa_key_mgmt & (WPA_KEY_MGMT_SAE | WPA_KEY_MGMT_FT_SAE)))
 		return false;
 
diff --git a/src/ap/wpa_auth_ie.c b/src/ap/wpa_auth_ie.c
index a4e86d019..455d0a89c 100644
--- a/src/ap/wpa_auth_ie.c
+++ b/src/ap/wpa_auth_ie.c
@@ -1619,7 +1619,7 @@ bool wpa_auth_write_fd_rsn_info(struct wpa_authenticator *wpa_auth,
 
 int wpa_auth_conf_gen_multibss_wpa_ie(struct wpa_auth_config *conf, u8 **wpa_ie, size_t *wpa_ie_len)
 {
-	u8 *pos, buf[WPA_IE_MAX_LEN] = {0};
+	u8 *pos, buf[WPA_IE_MAX_LEN + WPA_RSNO_IE_EXTRA_LEN] = {0};
 	int res = -1;
 	pos = buf;
 	if (conf->wpa & WPA_PROTO_RSN) {
@@ -1628,8 +1628,11 @@ int wpa_auth_conf_gen_multibss_wpa_ie(struct wpa_auth_config *conf, u8 **wpa_ie,
 		if (res < 0)
 			return res;
 		pos += res;
-		res = wpa_write_rsnxe(conf, pos,
-			buf + sizeof(buf) - pos);
+		if (conf->rsn_override_omit_rsnxe)
+			res = 0;
+		else
+			res = wpa_write_rsnxe(conf, pos,
+				buf + sizeof(buf) - pos);
 		if (res < 0)
 			return res;
 		pos += res;
@@ -1652,6 +1655,30 @@ int wpa_auth_conf_gen_multibss_wpa_ie(struct wpa_auth_config *conf, u8 **wpa_ie,
 		pos += res;
 	}
 
+	if ((conf->wpa & WPA_PROTO_RSN) && conf->rsn_override_key_mgmt) {
+		res = wpa_write_rsne_override(conf,
+			pos, buf + sizeof(buf) - pos);
+		if (res < 0)
+			return res;
+		pos += res;
+	}
+	if ((conf->wpa & WPA_PROTO_RSN) && conf->rsn_override_key_mgmt_2) {
+		res = wpa_write_rsne_override_2(conf, pos,
+			buf + sizeof(buf) - pos);
+		if (res < 0)
+			return res;
+		pos += res;
+	}
+	if ((conf->wpa & WPA_PROTO_RSN) &&
+	    (conf->rsn_override_key_mgmt ||
+	     conf->rsn_override_key_mgmt_2)) {
+		res = wpa_write_rsnxe_override(conf, pos,
+			buf + sizeof(buf) - pos);
+		if (res < 0)
+			return res;
+		pos += res;
+	}
+
 	if (*wpa_ie &&
 		((buf[0] == WLAN_EID_RSN) ||
 		(buf[0] == WLAN_EID_VENDOR_SPECIFIC))) {
diff --git a/src/common/wpa_common.h b/src/common/wpa_common.h
index c00d119f4..dcec082e6 100644
--- a/src/common/wpa_common.h
+++ b/src/common/wpa_common.h
@@ -32,6 +32,9 @@
 /* Max Security IE length RSNE/OSEN/RSNXE/WPA_VENDOR_IE */
 /* Max can be WPAIE(30) + RSNXE(3) */
 #define WPA_IE_MAX_LEN 33
+/* Additional buffer for RSN Overriding elements (RSNE Override, RSNE Override 2,
+ * RSNXE Override) when WPA3-Personal Compatibility Mode is enabled */
+#define WPA_RSNO_IE_EXTRA_LEN 120
 #endif /* CONFIG_VENDOR_MXL */
 
 #ifdef CONFIG_NO_TKIP
-- 
2.43.0

